Update dockerfile to PyMC v4

danhphan opened the pull request:
This PR updates the dockerfile to PyMC v4, which addresses some parts of the issue #5323
- Update the Dockerfile to a fresh base image for PyMC 4.0.0
- Fix NumPy `blas` warnings
- Run pm.model_to_graphviz(model) successfully

Todos:
- Update docs on how to build and run the docker image
- Create a GitHub Action to build & push the Docker image (need guidance) 

Thanks.

danhphan committed:
rename and update build docker container

danhphan committed:
update Dockerfile with pymc v4

danhphan committed:
add a script for running docker container from bash

danhphan committed:
update running docker container from juppyter notebook

danhphan commented:
The docker guideline could be somethings like:

### Build docker image

```
clone git@github.com:pymc-devs/pymc.git
cd pymc
bash scripts/build_container.sh
```
This building process may take a while. After the script successful finished, there should be a docker image named `pymc` in your machine. 
### How to run
You can now run the `pymc` docker container from `jupyter notebook` or from `bash`:

```
# Run with jupyter notebook
bash scripts/run_container_bash.sh
# Run with bash
bash scripts/run_container_bash.sh
```


The current docker document was commented in [CONTRIBUTING.md](https://github.com/pymc-devs/pymc/blob/main/CONTRIBUTING.md?plain=1). Should we update the docker guideline here in the CONTRIBUTING.md file? Or should it be moved into the documentation website like other parts (pull-request guideline, pre-commit guideline, ...), and then link to that docs.

cc @ricardoV94 @michaelosthege @twiecki @OriolAbril 

Thank you.

michaelosthege commented on scripts/Dockerfile:
```diff
@@ -1,18 +1,55 @@
+
+# For running from jupyter notebook
+EXPOSE 8888
+CMD ["conda", "run", "--no-capture-output", "-n", "pymc-dev-py39", "jupyter","notebook","--ip=0.0.0.0","--port=8888","--no-browser","--allow-root"]
```
Can `--allow-root` be removed, since the user was switched?

danhphan replied:
Yes agree, this makes sense. Let me remove it :)

michaelosthege reviewed (commented):
Looks great!

The CI pipeline will be an important addition, but let's do that on a follow-up PR.

OriolAbril commented:
> The current docker document was commented in [CONTRIBUTING.md](https://github.com/pymc-devs/pymc/blob/main/CONTRIBUTING.md?plain=1). Should we update the docker guideline here in the CONTRIBUTING.md file? Or should it be moved into the documentation website like other parts (pull-request guideline, pre-commit guideline, ...), and then link to that docs.

It should be on the docs, not on CONTRIBUTING.md (it could have a link there though if we think it is relevant enough)

canyon289 commented on scripts/Dockerfile:
```diff
@@ -1,18 +1,55 @@
 
-ADD $SRC_DIR /home/jovyan/
-RUN /bin/bash /home/jovyan/scripts/create_testenv.sh --global --no-setup
+USER root
```
Containers should not run as root. Is there a reason this is needed?

https://theorangeone.net/posts/containers-as-root/

danhphan replied:
Hi agree, it has been removed :)

canyon289 commented on scripts/run_container_bash.sh:
```diff
@@ -0,0 +1,12 @@
+#! /bin/bash
+
+SRC_DIR=${SRC_DIR:-`pwd`}
```
Can this script be merged with run_container_jupyter.sh? Theyre so similar it seems itd be easier to maintain one, rather than two

canyon289 replied:
Actually can all three bash scripts be merged? One with a nice CLI would be better for both users and maintenance

danhphan replied:
Yes, I have combine them into one script, and also update the Docker documents.

canyon289 commented:
Great stuff, thanks updating the dockerfile @danhphan!

danhphan committed:
remove root user from Dockerfile

danhphan committed:
simplify into one bash script

danhphan committed:
remove obsolete scripts

danhphan committed:
update documents on running PyMC in Docker

danhphan commented:
Hi, thank you all for the suggestions, I have:
- removed the root user and `--allow-root` from the Dockerfile,
- combined the three bash scripts into one, 
- updated the Docker documents in docs folder and also linked it in the [CONTRIBUTING.md](https://github.com/pymc-devs/pymc/blob/main/CONTRIBUTING.md?plain=1),
- removed obsolete scripts ([create_testenv.sh](https://github.com/pymc-devs/pymc/blob/main/scripts/create_testenv.sh) and [install_miniconda.sh](https://github.com/pymc-devs/pymc/blob/main/scripts/install_miniconda.sh)), but not sure if it is still used anywhere else?

OriolAbril commented on docs/source/contributing/docker_container.md:
```diff
@@ -0,0 +1,21 @@
+(pr_checklist)=
```
This is defining a target to which we can then link from anywhere in the docs to generate a link that points here. The `pr_checklist` label is already in use so it can't be used and it should be something specific to this file

danhphan replied:
Hi, got it.

OriolAbril commented on docs/source/contributing/docker_container.md:
````diff
@@ -0,0 +1,21 @@
+
+```bash
+$  bash scripts/docker_container.sh bash # running the container with bash
+$  bash scripts/docker_container.sh bash # running the container on jupyter notebook
````
I would remove the `$` so people can use the copy button that appears on code blocks directly. And both commands look the same, I think the 2nd one should use jupyter as argument or something of the sort.

OriolAbril commented on docs/source/contributing/docker_container.md:
````diff
@@ -0,0 +1,21 @@
+```bash
+$  bash scripts/docker_container.sh bash # logon to the container
+$  cd ./pymc/tests
+$  . ./../../scripts/test.sh # takes a while!
````
In addition to the dollar sign comment, do we use these `scripts/test.sh` in CI? Shouldn't the advise be to use pytest commands explicitly like we explain on https://www.pymc.io/projects/docs/en/stable/contributing/running_the_test_suite.html? We also recommend not running the whole test suite all at once unless necessary which that would seem to point against.

danhphan replied:
Yes, this content is from the old docs. I will remove them then. Also, not sure if the `scripts/test.sh` is used in any CI. If it is not used, we should also remove this script.

bwengals commented on scripts/Dockerfile:
```diff
@@ -1,18 +1,31 @@
+
+#Setup working forder
+RUN mkdir /home/jovyan/workspace
+WORKDIR /home/jovyan/workspace
```
I can double check myself tomorrow, but I *think* the jupyter base image makes a directory `/home/jupyter/work` that you're supposed to use as a mount target.  If so, I think you could remove 20-22.  (also very small spelling nit: forder -> folder on line 20)

danhphan replied:
Hi yes, I will check it as well :)

bwengals replied:
Yup just build the container, everything works great and mounts correctly.  There is both `work` and `workspace` in there though.  It's mentioned very briefly in their docs here: https://jupyter-docker-stacks.readthedocs.io/en/latest/using/common.html#additional-runtime-configurations.   It'd be a little neater to remove `RUN mkdir /home/jovyan/workspace` and use `WORKDIR /home/jovyan/work` instead.

Also, just a heads up that `jovyan` *should* be replaced with `$NB_USER` in your Dockerfile, since that is [something that's configurable too](https://jupyter-docker-stacks.readthedocs.io/en/latest/using/common.html#user-related-configurations).  I think it adds unnecessary complication to `scripts/docker_container.sh` though for this case, but might come in handy in other contexts.  Just want to mention if by chance you hadn't seen it!

bwengals commented on scripts/Dockerfile:
```diff
@@ -1,18 +1,31 @@
-FROM jupyter/minimal-notebook
+FROM jupyter/base-notebook:python-3.9.12
```
Good call on switching to `base-notebook`.  Any reason to fix the python version?  I think it might be preferable to have the container test build on "latest", just so no one needs to remember to bump versions.

danhphan replied:
Yes, using `base-notebook` saves some space. The image with a specific python 3.9 version will make it consistent to the python version in [environment-dev-py39.yml](https://github.com/pymc-devs/pymc/blob/main/conda-envs/environment-dev-py39.yml), which will be the only dev environment left after https://github.com/pymc-devs/pymc/pull/5842 will be merged. This image is also updated regularly on Dockerhub I think. 

It also work with latest `base-notebook`. I have tested in before, however the python version will be 3.10 (if I am not wrong), and our tests seem not ready with that yet https://github.com/pymc-devs/pymc/issues/5209 

I am happy to change it to the `latest` `base-notebook` image if you guys prefer this option.

michaelosthege replied:
we can also set up dependabot to automatically bump this version number when updates become available (I wanted open that PR once this one is merged)

bwengals replied:
Those are both good reasons to keep the version fixed, sounds great to me.

twiecki commented:
Is this good to merge?

canyon289 commented:
No, we're waiting on Bills comment about working directory

OriolAbril commented:
And my comments about docs still need to be addressed it looks like

bwengals commented on scripts/docker_container.sh:
```diff
@@ -0,0 +1,24 @@
+
+elif [[ $COMMAND = 'bash' ]]; then
+   docker run -it -v $SRC_DIR:/home/jovyan/workspace --rm  --name ${CONTAINER_NAME} ${CONTAINER_NAME} bash
+else
```
Might be nice to have an explicit elif for the `jupyter` option, not necessary though.  I also think it'd be nice to keep the bit at the end that [launches the notebook in the browser automatically](https://github.com/pymc-devs/pymc/blob/66215ace64b73f7c4ff2cc431ced9154e33da5e3/scripts/start_container.sh#L30-L32) but that's just a "nice to have" and could be a later PR.

danhphan replied:
Hi yes, the $COMMAND is set to `jupyter` as default in `COMMAND="${1:-jupyter}"`. So, after the docker image is built, we can run with `jupyter` as a default option.

bwengals commented:
I tried running it with all three options, `build`, `bash`, `jupyter`, everything works as expected!  Nicely done.  Other than the docs thing Oriol mentioned and the `workspace` -> `work` I think it's just about ready.

danhphan commented:
Hi, I have updated the docs and replaced `workspace` with `work`. Hopefully it is alright now :) Thank you

michaelosthege reviewed (approved):
Looks good to me.
There's no build or run script for Windows though, but we can also add that later.

danhphan commented:
Hi @michaelosthege , do you mean a script to create a Windows docker image? maybe we should not create a script for building a Windows-based docker image, since most docker image are linux-based, and most cloud systems use linux os.  In case of users that have Windows os with docker on their machine, they can also use the ubuntu image we have here to build and run the docker container. Thanks.

michaelosthege commented:
No I mean not a Windows Docker image, but a `docker_container.bat` with the same functionality as `docker_container.sh`.

But it's not critical since `docker_container.sh` is mostly a convenience thing.
The same can be achieved by modifying the commands below:

```bash
# Build the image, tagged as "pymc:local"
docker build -t pymc:local .

# Run in Jupyter notebook mode
docker run -it -p 8888:8888 -v %cd%:/home/jovyan/work --rm --name pymcJupyter pymc:local

# Run in interactive bash mode
docker run -it -v %cd%:/home/jovyan/work --rm --name pymcBash pymc:local bash
```

danhphan commented:
Yes, totally agree :)
